pkg/boot/kexec/purgatories.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2022 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:build ignore
+
+package main
+
+type asm struct {
+	name string
+	cc   []string
+	ld   []string
+	code string
+}
+
+var asms = []asm{
+	{
+		name: "to32bit_3000",
+		cc:   []string{"x86_64-linux-gnu-gcc", "-c", "-nostdlib", "-nostdinc", "-static"},
+		ld:   []string{"ld", "-N", "-e entry64", "-Ttext=0x3000"},
+		code: `
+# This code does its best to be PIC. As in, PIC by code, not by linker.
+# assumptions:
+# we have 8 bytes of stack to do ONE call 
+# we are in shared space -- code and data in dram
+# we are in long mode
+# in case of emergency, break glass.
+// 1: jmp 1b
+
+jmp 1f
+# save space for two arguments and stack
+.align 8
+entry: .long 0
+params: .long 0
+# This can be useful for being sure things are where we think they are,
+# and ensuring we are storing correctly.
+.long 0xdeadcafe, 0x12345678
+.long 0xcafe5555, 0xabcdefaa
+.align 128
+1:
+call 1f
+1:
+popq %rax
+andl $0xfffffff0, %eax
+movl %eax, %esp
+movl $gdt, %eax
+jmp 1f
+# recursive gdt. 
+	.balign 16
+gdt:	/* 0x00 unusable segment 
+	 * 0x08 unused
+	 * so use them as the gdt ptr
+	 */
+	.word	gdt_end - gdt - 1
+	.quad	gdt
+	.word	0, 0, 0
+			
+	/* Documented linux kernel segments */
+	/* 0x10 4GB flat code segment */
+	.word	0xFFFF, 0x0000, 0x9A00, 0x00CF
+	/* 0x18 4GB flat data segment */
+	.word	0xFFFF, 0x0000, 0x9200, 0x00CF
+gdt_end:
+# NOTE: NOT PIC
+to32indir:
+	.long to32
+	.long 0x10
+
+1: 
+	lgdt	gdt(%rip)
+	ljmp	*to32indir(%rip)
+to32:
+	.code32
+
+	.equ	CR0_PG,        0x80000000
+	/* Disable paging */
```

Comment:
https://github.com/u-root/u-root/blob/d29928f4a7dd41f32742fbd87fdb190fd5e2e885/pkg/boot/multiboot/internal/trampoline/trampoline_linux_amd64.s#L96

From here on, they are even the same.

Reply:
This code pattern is *everywhere*, although in this case I think the path was purgatory -> your trampoline .... And I'd still like  to wait just a bit until we're sure we don't need any bios calls; I'm no longer sure we even need to drop to 32 bit.